### ISSUE TITLE
build.sh: Separate dependency installation, compilation and installation of remill from each other

### ISSUE THREAD

**Issue by mewmew** [trailofbits/remill#93]
Felt very happy to discover the remill project today!

Running on Arch Linux, I had a quick look in the [AUR](https://aur.archlinux.org/packages/?O=0&K=remill) for a remill package, but was unable to locate one.

When running the `build.sh` script, the following error was encountered.

```
[u@x61s remill]$ ./build.sh 
[+] Creating '/home/u/Desktop/remill/build'
~/Desktop/remill/build ~/Desktop/remill
[+] Downloading Clang+LLVM..
--2017-02-14 22:21:11--  http://llvm.org/releases/3.9.0/clang+llvm-3.9.0-x86_64-linux-gnu-arch-rolling.tar.xz
Resolving llvm.org... 192.17.58.186
Connecting to llvm.org|192.17.58.186|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: http://releases.llvm.org/3.9.0/clang+llvm-3.9.0-x86_64-linux-gnu-arch-rolling.tar.xz [following]
--2017-02-14 22:21:11--  http://releases.llvm.org/3.9.0/clang+llvm-3.9.0-x86_64-linux-gnu-arch-rolling.tar.xz
Resolving releases.llvm.org... 192.17.58.186
Reusing existing connection to llvm.org:80.
HTTP request sent, awaiting response... 404 Not Found
2017-02-14 22:21:12 ERROR 404: Not Found.
```

Note, the system has both llvm and clang version 3.9.1 installed.

```
u@x61s ~> pacman -Q llvm clang
llvm 3.9.1-2
clang 3.9.1-2
```

**Comment by mewmew** [comment]
I've started looking into how one may go about bringing remill to Arch Linux.

An initial AUR package for remill is located at https://aur.archlinux.org/packages/remill-git/

There are still some issues to resolve. For instance, from what I can tell a lot of the build scripts and test cases use hard-coded paths. Is there a way to specify to `build.sh` to not build LLVM and Clang from source if they exist on the system (and are of the right version, 3.9.x)? For now, I've pulled the parts from the `build.sh` script which seem relevant if you have these tools and libraries already installed on your system.

From the [PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=remill-git#n42) of the remill-git AUR package.
```bash
  #
  # BUILD HERE
  #
  echo "[+] Compiling protobufs"
  ./scripts/unix/compile_protobufs.sh
  echo "[+] Running cmake"
  cmake -DCMAKE_INSTALL_PREFIX:PATH="$pkgdir/usr" -DLLVM_DIR=/usr/include/llvm -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DCMAKE_LLVM_LINK=/usr/bin/llvm-link .
  echo "[+] Building semantics"
  make semantics
  echo "[+] Building remill"
  make all
```

I hope to be able to replace those commands with `./build.sh` in the future, so that the AUR package will stay up to date as the dependencies and build instructions of remill changes.

**Comment by mewmew** [comment]
Note, the references of hard-coded paths in the test cases specifically refers to the `CXX` variable in [scripts/x86/generate_tests.sh](https://github.com/trailofbits/remill/blob/master/scripts/x86/generate_tests.sh#L22), which assumes that `clang++` is always located in the build directory.

```bash
CXX=$DIR/build/llvm/bin/clang++
```

If the system copy of `clang++` is of the right version, it would be lovely to support using the system tool instead of depending on building LLVM and Clang each time. Note that this has a drastic implication on the compile time of remill, for new users who wants to get up and running.

With LLVM and Clang it may takes hours, while without it is a matter of minutes. At least on my hardware.

Cheers /u

**Comment by pgoodman** [comment]
So, the problem is that there is no official pre-built binaries for LLVM and Clang 3.9 on Arch Linux, and so you'd need to download and compile LLVM from source?

**Comment by mewmew** [comment]
> So, the problem is that there is no official pre-built binaries for LLVM and Clang 3.9 on Arch Linux, and so you'd need to download and compile LLVM from source?

Sorry for the confusing explanation. It is rather quite the opposite. On Arch there are pre-built binaries for LLVM and Clang 3.9, and these are distributed through the Arch Linux package management system, and not from llvm.org directly. There only seem to be pre-build binaries for Ubuntu, Debian, openSUSE and Fedora released directly from http://releases.llvm.org/3.9.0/.

My thoughts were to use the pre-built binaries for LLVM and Clang 3.9 directly from the package management system of Arch Linux, as these are well supported on the Arch playform.

The main problem is that the pbuild.sh` script expects the LLVM and Clang binaries to be located in the `build/` directory of remill, and it does not seem possible to specify using the system binaries from what I can tell.

From [build.sh](https://github.com/trailofbits/remill/blob/master/build.sh#L63)
```bash
LLVM_DIR=${BUILD_DIR}/${LLVM_DIR}
```

Note, this issue is related to the other dependencies of remill as well (e.g. gtest, ...), since the three distinct tasks dependency installation, compilation and installation of remill are all bundled tightly together in the `build.sh` script. If we were to separate dependency installation, compilation and installation of remill into three scripts, or three rules in a Makefile, I'd be happy.

That is, to create a `install_deps.sh` (or something similar), containing:

```bash
...
if [ ! -d ${LLVM_DIR} ]; then
  echo "[+] Extracting.."
  mkdir ${LLVM_DIR}
  tar xf ${FILE} -C ${LLVM_DIR} --strip-components=1 
fi
echo "[+] Installing Xed"
$DIR/scripts/unix/install_xed.sh
echo "[+] Installing gtest"
$DIR/scripts/unix/install_gtest.sh
echo "[+] Compiling protobufs"
$DIR/scripts/unix/compile_protobufs.sh
echo "[+] Running cmake"
```

And a `build.sh` containing, where the path of LLVM can either be specified or inferred by cmake (as is already the case with [gflags](https://github.com/trailofbits/remill/blob/master/cmake/FindGFLAGS.cmake), glog and xed):

```bash
cmake -DLLVM_DIR=${LLVM_DIR} -DCMAKE_C_COMPILER=${BINDIR}/clang -DCMAKE_CXX_COMPILER=${BINDIR}/clang++ -DCMAKE_LLVM_LINK=${BINDIR}/llvm-link ${DIR}
echo "[+] Building semantics"
make semantics
echo "[+] Building remill"
make all
```

And `install.sh`:

```bash
echo "[+] Installing"
sudo make install
```

To sum up, split `build.sh` into three files (or three rules of a Makefile), `install_deps.sh`, `build.sh` and `install.sh`.

I'll update the topic of this issue accordingly.

**Comment by pgoodman** [comment]
So, what are the next steps from here? Is there stuff that should be submitted as a PR?

**Comment by mewmew** [comment]
> So, what are the next steps from here? Is there stuff that should be submitted as a PR?

Hej Peter,

Sorry for a late reply. I've been quite busy as of late.

The next step would be to break build.sh into separate shell scripts, one for installing dependencies (which could be ignored on systems that install those dependencies from other sources) and one for building remill.

Anyone may feel free to take a stab at submitting a PR for this. I shall have more time on my hands in a couple of months. If it has not be patched by then, I may give it a try. Hopefully a new version of IDA may have arrived by then, which would solve some of the issues related to 32-bit Python running on a 64-bit system that I've been experiencing; and that have effectively limited me from playing with remill as much as I would have like.

I was glad to meet you last month Peter. Wish you a happy spring : )

Cheers /u

**Comment by pgoodman** [comment]
Good news, @alessandrogario is working on some better dependency management that might make Arch Linux support easier :-D This will have a nice complimentary thing for McSema!

Happy Spring! The weather has been really nice here recently :-)

**Comment by mewmew** [comment]
Oh, lovely to hear! I'll await the update with anticipation : )

Happy spring! We've also had most tremendously lovely weather, perfect for skiing in the North!!

kram /u

**Comment by alessandrogario** [comment]
Hello @mewmew !

You can track my progress my checking out the alessandro/refactor/cmake branch. It should already be working (and I've also added a PKGBUILD script to build an ArchLinux package). Easiest way to build my branch is to use the script under remill/library_repository_installer.

Speaking about the IDAPython issue, I'm also using IDA on a 32-bit linux machine. I made a script to recompile python for IDA, and you can find it in the mcsema repository under "tools/build_idapython.sh".

As a last note, there's a type conversion error in the code right now, so you may have to fix it (or ignore it by removing the -Wconversion line from remill/Arch/X86/Runtime/CMakeLists.txt)

**Comment by mewmew** [comment]
@alessandrogario Oh, wow! Thanks a lot for working on this!

I'll definitely take a look at both your remill PKGBUILD and the library_repository_installer script.

Also lovely to hear that you've made a script to build a python version for IDA on 64-bit machines.

You officially made my day Alessandro!

Cheerful regards /u

**Comment by mewmew** [comment]
@alessandrogario I found the `library_repository_installer` directory. However, I did not manage to locate a PKGBUILD. Did you upload it somewhere?

I've added a preliminary (non-working) [remill-git package](https://aur.archlinux.org/packages/remill-git/) to the AUR. If you have a working PKGBUILD, feel free to push it to the AUR. Let me know if I need to grant you access somehow, since I was the first to create the AUR package.

Cheers /u

**Comment by mewmew** [comment]
With the release of IDA 7.0, I really wanted to take remill (and MC-Semantics) out for a spin. @alessandrogario Is the PKGBUILD working on your machine?

I updated the source as follows,

```diff
u@x220 ~/D/r/d/ArchLinux> git diff .
diff --git a/distro_packages/ArchLinux/PKGBUILD b/distro_packages/ArchLinux/PKGBUILD
index 83ff838..3e2cfa4 100644
--- a/distro_packages/ArchLinux/PKGBUILD
+++ b/distro_packages/ArchLinux/PKGBUILD
@@ -12,7 +12,7 @@ makedepends=('git')
 provides=('remill')
 conflicts=('remill')
 
-source=("${pkgname}::git+https://github.com/trailofbits/remill.git#branch=alessandro/refactor/cmake")
+source=("${pkgname}::git+https://github.com/trailofbits/remill.git#branch=archive/alessandro/refactor/cmake")
 sha1sums=('SKIP')
 
 pkgver() {
```

When running, it fails with the following error:

```
u@x220 ~/D/r/d/ArchLinux> makepkg
==> Making package: remill 20170418_77bb5f2-1 (Sun Sep 17 02:06:19 CEST 2017)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Retrieving sources...
  -> Updating remill git repo...
Fetching origin
==> Validating source files with sha1sums...
    remill ... Skipped
==> Extracting sources...
  -> Creating working copy of cmake git repo...
Reset branch 'makepkg'
==> Starting pkgver()...
==> Removing existing $pkgdir/ directory...
==> Starting build()...
make: *** No rule to make target 'clean'.  Stop.
==> ERROR: A failure occurred in build().
    Aborting...
```

Any ideas?

Cheers /u

**Comment by pgoodman** [comment]
So we have most dependencies factored into cxx-common. I am not sure what the state of that cmake branch is, its features may have since been subsumed. @alessandrogario, can you comment on this?

**Comment by alessandrogario** [comment]
Hello @mewmew ,

I've updated the PKGBUILD script, do you mind trying it out? It's in the alessandro/feature/archlinux-package branch.

Thanks!

**Comment by mewmew** [comment]
Hej @alessandrogario,

I tried to compile the AUR package, and it worked well. Great job!

However, when installing it on the system I got the following error:

```
$ sudo pacman -U remill-20171201_bfdf3c7-1-x86_64.pkg.tar.xz 
loading packages...
resolving dependencies...
looking for conflicting packages...

Packages (1) remill-20171201_bfdf3c7-1

Total Installed Size:  37.77 MiB

:: Proceed with installation? [Y/n] 
(1/1) checking keys in keyring                                                                                                                  [########################################################################################] 100%
(1/1) checking package integrity                                                                                                                [########################################################################################] 100%
(1/1) loading package files                                                                                                                     [########################################################################################] 100%
(1/1) checking for file conflicts                                                                                                               [########################################################################################] 100%
error: failed to commit transaction (conflicting files)
remill: /bin exists in filesystem
Errors occurred, no packages were upgraded.
```

I think the binaries should be placed in `/usr/bin` rather than `/bin`.

**Comment by alessandrogario** [comment]
Hello @mewmew ,

you are right! I have made a mistake in setting the CMAKE_INSTALL_PREFIX and forgot about DESTDIR in the install step. I have now pushed a fix in my branch, can you please have a look at it?

Thanks!

**Comment by mewmew** [comment]
Now it works! Thanks @alessandrogario :)

```
==> Finished making: remill 20171206_ac62ffc-1 (Thu Dec  7 00:40:36 CET 2017)
```

**Comment by mewmew** [comment]
@alessandrogario Would you like for me to disown the [remill-git](https://aur.archlinux.org/packages/remill-git/) package at the AUR, so that you could put your PKGBUILD there?

**Comment by alessandrogario** [comment]
Hey @mewmew !
I already have two packages on AUR I am no longer maintaining! You should keep it (and you can also update it... :D )

**Comment by mewmew** [comment]
Ok, I'll update it to include your PKGBUILD if that's ok with you?

**Comment by alessandrogario** [comment]
Yes of course! Feel free to change it as you see fit and send us back a PR!

**Comment by mewmew** [comment]
Lovely. I've updated it now. Would it be possible to merge the alessandro/feature/archlinux-package branch into master?

I'd like to update the AUR package to use the master branch,

```diff
-source=("${pkgname}::git+https://github.com/trailofbits/remill.git#branch=alessandro/feature/archlinux-package")
+source=("${pkgname}::git+https://github.com/trailofbits/remill.git")
```

But that gives the following error when building: 

```
CMake Error at CMakeLists.txt:250 (message):
  InstallExternalTarget: The following path does not exists:
  /home/u/Desktop/remill_play/remill-git/src/libraries/llvm/bin/clang-3.9
Call Stack (most recent call first):
  CMakeLists.txt:321 (InstallExternalTarget)


-- Configuring incomplete, errors occurred!
See also "/home/u/Desktop/remill_play/remill-git/src/build/CMakeFiles/CMakeOutput.log".
See also "/home/u/Desktop/remill_play/remill-git/src/build/CMakeFiles/CMakeError.log".
==> ERROR: A failure occurred in build().
    Aborting...
```

Also minor note, but if you happen to update the PKGBUILD file, also update the pkgname to `remill-git` as that is the convention for packages relying on git instead of version numbers.

```diff
-pkgname=remill
+pkgname=remill-git
```

Cheers!
/u

**Comment by alessandrogario** [comment]
Hey @mewmew ,

I have had to update the CMakeLists.txt a little, so it's normal it doesn't work right now. I have made a PR, and also renamed the package to remill-git as you suggested. Once merged, it will work with the master branch!

**Comment by mewmew** [comment]
Perfect!

You may feel free to close this issue, one that PR has been merged.
